### PR TITLE
Create ActionManager component and containers for device detail page 

### PR DESCRIPTION
**Pull request by ajaypratap003**
### Type of change

- Enhancement / new feature
- Refactoring

### Description
1. Create action, reducer function for saving device action 
2. Create ActionManager component to handle all the device detail page's actions
3. Create Edit metadata container 
4. Move MetaData component in separate components folder
5. Fixed  some warnings 
6. Create functions/events in device details page

image

image

### Checklist

- [ ] Update/write design documentation in `./documentation/design`
- [x] Edit tests and make sure they pass
- [ ] Resolve all warnings
- [x] Update license
- [ ] Write test cases for null checks in Unit Tests
- [ ] Reference relevant issue(s) and close them after merging

### CODE REVIEW

File: console/console-init/ui/src/modules/iot-device-detail/components/DeviceInfo/DeviceInfo.tsx [hunk 1]
```diff
@@ -122,7 +146,7 @@ export const DeviceInfo: React.FC<IDeviceInfoProps> = ({
                           </GridItem>
                         );
                       })}
-                    {!(deviceList && deviceList.length > 0) && (
+                    {deviceList?.length <= 0 && (
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device-detail/components/DeviceInfo/DeviceInfo.tsx:149]: if deviceList is undefined then in that case we should display the same Text

File: console/console-init/ui/src/stories/MetadataListTable.stories.tsx [hunk 2]
```diff
@@ -42,7 +42,7 @@ export const MetadataTable = () => (
       id={"metadata-table"}
       dataList={metadataList}
       aria-label={"metadata table"}
-      aria-labelledby-header={"medata data header"}
+      aria-labelledby={"medata data header"}
```
Comment (anukritijha) [inline console/console-init/ui/src/stories/MetadataListTable.stories.tsx:45]: Typo "metadata"

**Comment by ajaypratap003** [comment]
> Came across this error while running the project:
> image

This issue has fixed.

**Comment by ajaypratap003** [comment]
> For device with deviceId 12. The site breaks if we click `Delete Credentials`
> 
> Screenshot from 2020-06-15 16-16-39

@rkpattnaik780 , mock server is not working as expected. Need to discuss with backend team.

**Comment by rkpattnaik780** [comment]
> > For device with deviceId 12. The site breaks if we click `Delete Credentials`
> > Screenshot from 2020-06-15 16-16-39
> 
> @rkpattnaik780 , mock server is not working as expected. Need to discuss with backend team.

Okay!

File: console/console-init/ui/src/modules/iot-device/containers/EditGatewaysContainer/EditGatewaysContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  return (
+    <>
+      <AddGateways
+        gateways={gatewayList}
+        header={`Edit gateways of device ${deviceid}`}
+        returnGateways={getGateways}
+      />
+      <Flex className={styles.button_padding}>
+        <FlexItem>
+          <Button variant={ButtonVariant.primary} onClick={onSave}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device/containers/EditGatewaysContainer/EditGatewaysContainer.tsx:67]: please add Id on button components.

File: console/console-init/ui/src/modules/iot-device/containers/EditGatewaysContainer/EditGatewaysContainer.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            Save
+          </Button>
+        </FlexItem>
+        <FlexItem>
+          <Button variant={ButtonVariant.link} onClick={onCancel}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device/containers/EditGatewaysContainer/EditGatewaysContainer.tsx:72]: please add id on button component

File: console/console-init/ui/src/modules/iot-device/containers/EditDeviceInJsonContainer/EditDeviceInJsonContainer.tsx [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      <JsonEditor
+        readOnly={false}
+        value={JSON.stringify(getDeviceJson(), undefined, 2)}
+        className={styles.editor_border}
+      />
+      <br />
+      <br />
+      <Flex>
+        <FlexItem>
+          <Button variant={ButtonVariant.primary} onClick={onSave}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device/containers/EditDeviceInJsonContainer/EditDeviceInJsonContainer.tsx:78]: please add id on button component

File: console/console-init/ui/src/modules/iot-device/containers/EditDeviceInJsonContainer/EditDeviceInJsonContainer.tsx [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            Save
+          </Button>
+        </FlexItem>
+        <FlexItem>
+          <Button variant={ButtonVariant.secondary} onClick={onCancel}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device/containers/EditDeviceInJsonContainer/EditDeviceInJsonContainer.tsx:83]: please add id on button component

File: console/console-init/ui/src/modules/iot-device/containers/EditCredentialsContainer/EditCredentialsContainer.tsx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  return (
+    <>
+      <Title size={"2xl"}>Edit credentials</Title>
+      <br />
+      <AddCredential />
+      <br />
+      <br />
+      <Flex>
+        <FlexItem>
+          <Button variant={ButtonVariant.primary} onClick={onSave}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device/containers/EditCredentialsContainer/EditCredentialsContainer.tsx:47]: please add id on button component

File: console/console-init/ui/src/modules/iot-device/containers/EditCredentialsContainer/EditCredentialsContainer.tsx [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            Save
+          </Button>
+        </FlexItem>
+        <FlexItem>
+          <Button variant={ButtonVariant.secondary} onClick={onCancel}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device/containers/EditCredentialsContainer/EditCredentialsContainer.tsx:52]: please add id on button component

File: console/console-init/ui/src/modules/iot-device/containers/AddGatewaysContainer/AddGatewaysContainer.tsx [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+  return (
+    <>
+      <AddGateways
+        header={`Add gateways of device ${deviceid}`}
+        returnGateways={getGateways}
+      />
+      <Flex className={styles.button_padding}>
+        <FlexItem>
+          <Button variant={ButtonVariant.primary} onClick={onSave}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device/containers/AddGatewaysContainer/AddGatewaysContainer.tsx:52]: please add id on button component

File: console/console-init/ui/src/modules/iot-device/containers/AddGatewaysContainer/AddGatewaysContainer.tsx [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+            Save
+          </Button>
+        </FlexItem>
+        <FlexItem>
+          <Button variant={ButtonVariant.link} onClick={onCancel}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device/containers/AddGatewaysContainer/AddGatewaysContainer.tsx:57]: please add id on button component

File: console/console-init/ui/src/modules/iot-device/containers/AddCredentialsContainer/AddCredentialsContainer.tsx [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  return (
+    <>
+      <Title size={"2xl"}>Add credentials</Title>
+      <br />
+      <AddCredential />
+      <br />
+      <br />
+      <Flex>
+        <FlexItem>
+          <Button variant={ButtonVariant.primary} onClick={onSave}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device/containers/AddCredentialsContainer/AddCredentialsContainer.tsx:44]: please add id on button component

File: console/console-init/ui/src/modules/iot-device/containers/AddCredentialsContainer/AddCredentialsContainer.tsx [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+            Save
+          </Button>
+        </FlexItem>
+        <FlexItem>
+          <Button variant={ButtonVariant.secondary} onClick={onCancel}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device/containers/AddCredentialsContainer/AddCredentialsContainer.tsx:49]: please add id on button component

File: console/console-init/ui/src/modules/iot-device-detail/components/DeviceInfo/ErrorStateAlert.tsx [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+          </Button>
+        </>
+      );
+    } else if (ErrorState.CONFLICTING) {
+      return (
+        <>
+          This device has two connection types that are conflicting with each
+          other. You can choose actions to fix this issue.
+          <br />
+          <Button variant={ButtonVariant.link} onClick={deleteGateways}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device-detail/components/DeviceInfo/ErrorStateAlert.tsx:66]: please add id on button component

File: console/console-init/ui/src/modules/iot-device-detail/components/DeviceInfo/ErrorStateAlert.tsx [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+            Delete gateways
+          </Button>{" "}
+          or
+          <Button variant={ButtonVariant.link} onClick={deleteCredentials}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device-detail/components/DeviceInfo/ErrorStateAlert.tsx:70]: please add id on button component

File: console/console-init/ui/src/components/UpdatePassword/UpdatePassword.tsx [hunk 15]
```diff
@@ -98,13 +106,9 @@ export const UpdatePassword = () => {
...
       <Alert variant="info" isInline title="Want to update the secrets?">
         If you want to update the secrets, you can go to the
-        {/**
-         * TODO: add link path
-         */}
-        <NavLink id="update-password-edit-credentials-navlink" to="/">
-          {" "}
-          Edit Credentials{" "}
-        </NavLink>
+        <Button variant={"link"} onClick={onClickEditCredentials}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/components/UpdatePassword/UpdatePassword.tsx:109]: please add id on button component

File: console/console-init/ui/src/modules/iot-device/containers/EditMetadataContainer/EditMetadataContainer.tsx [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+        <br />
+        <Grid gutter="sm">
+          <GridItem span={6}>
+            <CreateMetadata />
+          </GridItem>
+        </Grid>
+        <br />
+        <Flex>
+          <FlexItem>
+            <Button variant={ButtonVariant.primary} onClick={onSave}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device/containers/EditMetadataContainer/EditMetadataContainer.tsx:66]: please add id on button component

File: console/console-init/ui/src/modules/iot-device/containers/EditMetadataContainer/EditMetadataContainer.tsx [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+              Save
+            </Button>
+          </FlexItem>
+          <FlexItem>
+            <Button variant={ButtonVariant.link} onClick={onCancel}>
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device/containers/EditMetadataContainer/EditMetadataContainer.tsx:71]: please add id on button component

**Pull request merged**